### PR TITLE
Document user-selectable C macros

### PR DESCRIPTION
**Pull request by da-woods**
I've shown the 4 that are most user-relevant clearly, and hidden most of the rest by default (but documented them if people really want to read them).

### CODE REVIEW

File: docs/src/userguide/source_files_and_compilation.rst [hunk 1]
```diff
@@ -1123,3 +1123,115 @@ argument to ``cythonize``::
...
+
+``CYTHON_USE_TYPE_SPECS``
+    Defines ``cdef classes`` as `"heap types" <https://docs.python.org/3/c-api/typeobj.html#heap-types>`_
+    rather than "static types".  Practically this does not change a lot from a user
+    point of view, but it is needed to implement Limited API support.
+    
+There are a further list of macros which turn off various optimizations or language
+features.  Under normal circumstance Cython enables these automatically based on the
+version of Python you are building a module for so do not be tempted to use them
+to try to enable extra optimizations - supported optimizations are enabled by
```
Comment (da-woods) [inline docs/src/userguide/source_files_and_compilation.rst:1171]: I've really tried to make it clear that there's no benefit to users going an manually enabling them - the danger seems to be that people think there's secret optimizations here that they as an "advanced user" should be able to access.  I suspect there isn't a font size big enough to get this message across but it's worth a try.

**Comment by matusvalo** [comment]
I think `CYTHON_EXTERN_C` macro is missing. It is mentioned elsewhere https://cython.readthedocs.io/en/latest/src/userguide/external_C_code.html#c-public-declarations but I suspect it should be mentioned here too...

File: docs/src/userguide/source_files_and_compilation.rst (hunk 1 above)
Comment (matusvalo) [reply docs/src/userguide/source_files_and_compilation.rst:1171]: > I suspect there isn't a font size big enough

Maybe putting it in warning block will help...

**Comment by da-woods** [comment]
> I think `CYTHON_EXTERN_C` macro is missing.

Added. Thanks.

> Maybe putting it in warning block will help...

Currently I'm trying not to draw attention to the section - it isn't really pitfall or warning - just something that doesn't apply to most people

File: docs/src/userguide/source_files_and_compilation.rst [hunk 2]
```diff
@@ -1123,3 +1123,119 @@ argument to ``cythonize``::
...
+            
+        ``CYTHON_USE_DICT_VERSIONS``
+            Tries to optimize attribute lookup by using versioned dictionaries
+            where supported.
+            
+        ``CYTHON_USE_EXC_INFO_STACK``
+            Uses an internal structure to track exception state.
+            
+        ``CYTHON_UPDATE_DESCRIPTOR_DOC``
+            Attempts to provide docstrings for special methods.
```
Comment (scoder) [inline docs/src/userguide/source_files_and_compilation.rst:1241]:
```suggestion
        ``CYTHON_USE_DICT_VERSIONS``
            Try to optimize attribute lookup by using versioned dictionaries
            where supported.
            
        ``CYTHON_USE_EXC_INFO_STACK``
            Use an internal structure to track exception state,
            used in CPython 3.7 and later.
            
        ``CYTHON_UPDATE_DESCRIPTOR_DOC``
            Attempt to provide docstrings also for special (double underscore) methods.
```

File: docs/src/userguide/source_files_and_compilation.rst [hunk 3]
```diff
@@ -1123,3 +1123,119 @@ argument to ``cythonize``::
...
+        
+        ``CYTHON_USE_TYPE_SLOTS``
+            If enabled Cython will directly access members of the ``PyTypeObject``
+            struct.
+            
+        ``CYTHON_USE_PYTYPE_LOOKUP``
+            More efficient access into properties of C classes.
+            
+        ``CYTHON_USE_ASYNC_SLOTS``
+            Support for the ``tp_as_async`` attribute on type objects.
```
Comment (scoder) [inline docs/src/userguide/source_files_and_compilation.rst:1194]:
```suggestion
            Support the ``tp_as_async`` attribute on type objects.
```

File: docs/src/userguide/source_files_and_compilation.rst [hunk 4]
```diff
@@ -1123,3 +1123,119 @@ argument to ``cythonize``::
...
+them to 0 to *disable* optimizations.  They are listed below for completeness but
+hidden by default since most users will be uninterested in changing them.
+
+.. tabs::
+    .. tab:: Hide
+    
+    .. tab:: Show
+        
+        ``CYTHON_USE_TYPE_SLOTS``
+            If enabled Cython will directly access members of the ``PyTypeObject``
```
Comment (scoder) [inline docs/src/userguide/source_files_and_compilation.rst:1187]:
```suggestion
            If enabled, Cython will directly access members of the ``PyTypeObject``
```

File: docs/src/userguide/source_files_and_compilation.rst [hunk 5]
```diff
@@ -1123,3 +1123,119 @@ argument to ``cythonize``::
...
+            struct.
+            
+        ``CYTHON_USE_PYTYPE_LOOKUP``
+            More efficient access into properties of C classes.
```
Comment (scoder) [inline docs/src/userguide/source_files_and_compilation.rst:1191]:
```suggestion
            Use the internal `_PyType_Lookup()` function for more efficient access
            to properties of C classes.
```

**Review by scoder — COMMENTED** [review]
Let's try to use a bit more consistent wording. And link to PEPs that we mention.

**Comment by scoder** [comment]
Very good idea, nice to have them listed in the docs.

**Pull request merged**